### ISSUE TITLE
Skip invariant overload arguments that are Any/Unknown

### ISSUE THREAD

**Issue by twoertwein** [microsoft/pyright#4546]
**Is your feature request related to a problem? Please describe.**
Most arguments are the same type across overloads. If an argument that has the same type across all overloads, is Any/Unknown at call-site, pyright seems to pick the last overload.

```py
from __future__ import annotations
from typing import overload, Literal, Any

@overload
def test(x: str, *, flag: Literal[True]) -> int:
    ...

@overload
def test(x: str, *, flag: Literal[False] = ...) -> str:
    ...

@overload
def test(x: str, *, flag: bool = ...) -> int | str:
    ...

def test(x: str, *, flag: bool = False) -> int | str:
    ...

reveal_type(test("0")) # mypy/pyright: str
reveal_type(test("0", flag=True)) # mypy/pyright: int
reveal_type(test("0", flag=False)) # mypy/pyright: str

def unknown_any() -> Any:
    ...

reveal_type(test(unknown_any(), flag=False)) # mypy: str, pyright: str | int
```

**Describe the solution you'd like**
If an argument is Any/Unknown at call-site and the overloads have the same types for it, it might be nice to pick the overload based on only the remaining arguments.

**Comment by erictraut** [comment]
Here's the algorithm that pyright uses currently:
1. It first tries to match the provided arguments to parameters in all of the overloads by arity (i.e. the count of the parameters). It eliminates any overloads where the arity doesn't match.
2. It then evaluates the types of the arguments and matches them against the types of the remaining overload parameters.
3. If none of them match, it performs union expansion on any arguments whose types evaluate to unions.
4. If multiple overloads match due to one or more arguments evaluating to `Any` or `Unknown`, it evaluates the call expression type result as the union of all return types for all matching overloads.

Rule 4 is what applies here. Since the argument for `x` is `Any` and the argument for `flag` is `False`, there are two overloads that match (the second and third), so the result is a union of their respective return types (`str` and `int | str`).

I thought this behavior was consistent with mypy's behavior, but evidently it's subtly different. I'll need to investigate further.

**Comment by erictraut** [comment]
Pyright isn't picking the last call site here. It's doing something much more sophisticated. It's actually picking the second and third call sites and noticing that one or more of the arguments is `Any`. Since the result is ambiguous, it then looks at the respective return types of the remaining matches and determines that they return `str` and `int | str`. Since the former is a subtype of the latter, it chooses the wider of the two types. If there were no overlap between the remaining return types, it would return `Unknown`.

I thought this behavior was consistent with mypy's. Here's what mypy's documentation says:

> However, there are two exceptions to the “pick the first match” rule. First, if multiple variants match due to an argument being of type Any, mypy will make the inferred type also be Any.

Evidently, mypy is doing something subtly different here. I'll investigate further.

It's unfortunate that mypy chose to implement such baroque overload rules. It makes it very difficult to get consistent overload behavior across type checkers. I much prefer TypeScript's overload behavior where it always picks the first matching overload with no union expansion. That's very straightforward to document, understand, validate, and implement. Alas.

**Comment by erictraut** [comment]
I've updated pyright's overload matching logic to more closely match that of mypy so they produce the same result in this case. This will be included in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.292, which I just published. It will also be included in a future release of pylance.
